Adding async reset info (#25)
chipsalliance/firrtl-spec: spec.md

Clock
```

### Reset Type

The abstract `Reset`{.firrtl} type is either inferred to `UInt<1>`{.firrtl}
(synchronous reset) or `AsyncReset`{.firrtl} (asynchronous reset) during
compilation.

Synchronous resets used in registers will be mapped to a hardware description
language representation for synchronous resets.

> The following example demonstrates usage of a synchronous reset.

Reviewer: ```suggestion
The following example shows an uninferred reset which can/must/will be lowered to a `UInt<1>`{.firrtl} or `AsyncReset`{.firrtl}.
```

``` firrtl
input clock : Clock
input r : UInt<1>
input x : UInt<8>
wire reset: Reset
reset <= r
reg y : UInt<8>, clock with : (reset => (reset, UInt(123)))
; ...
```

> 

Reviewer: ```suggestion
The following shows a synchronous reset...
```

show the thing with type `input reset : UInt<1>`

Asynchronous resets used in registers will be mapped to a hardware description
language representation for asynchronous resets.

The following example demonstrates usage of an asynchronous reset.

``` firrtl
input clock : Clock
input reset : AsyncReset
input x : UInt<8>
reg y : UInt<8>, clock with : (reset => (reset, UInt(123)))
; ...
```

Inference rules are as follows:

> 1. An abstract reset driven by and/or driving only asynchronous resets will be

Reviewer: let's pick a consistent term of either `abstract` or `uninferred`

inferred as asynchronous reset
1. An abstract reset driven by and/or driving both asynchronous and synchronous
resets will error
1. Otherwise, the reset is inferred as synchronous (i.e. the abstract reset is
only invalidated or is driven by or drives only synchronous resets)

> Note that an exception will be thrown if a `Reset`{.firrtl} is driven by

Reviewer: combine with line 361

Clock
```

### Reset Type

> The abstract `Reset`{.firrtl} type is either inferred to `UInt<1>`{.firrtl}
> (synchronous reset) or `AsyncReset`{.firrtl} (asynchronous reset) after
> compilation.

Reviewer 2: Nit:
```suggestion
The abstract `Reset`{.firrtl} type is either inferred to `UInt<1>`{.firrtl}
(synchronous reset) or `AsyncReset`{.firrtl} (asynchronous reset) after
compilation.
```
```suggestion
The abstract `Reset`{.firrtl} type is either inferred to `UInt<1>`{.firrtl}
(synchronous reset) or `AsyncReset`{.firrtl} (asynchronous reset) during
compilation.
```

language representation for synchronous resets.

The following example demonstrates usage of a synchronous reset.

``` firrtl
input clock : Clock
input r : UInt<1>
input x : UInt<8>
wire reset: Reset
reset <= r
reg y : UInt<8>, clock with : (reset => (reset, UInt(123)))
; ...
```

Asynchronous resets used in registers will be mapped to a hardware description
language representation for asynchronous resets.

The following example demonstrates usage of an asynchronous reset.

``` firrtl
input clock : Clock
input reset : AsyncReset
input x : UInt<8>
reg y : UInt<8>, clock with : (reset => (reset, UInt(123)))
; ...
```

Inference rules are as follows:

1. An abstract reset driven by and/or driving only asynchronous resets will be
inferred as asynchronous reset
1. An abstract reset driven by and/or driving both asynchronous and synchronous
resets will error
1. Otherwise, the reset is inferred as synchronous (i.e. the abstract reset is
only invalidated or is driven by or drives only synchronous resets)

Note that an exception will be thrown if a `Reset`{.firrtl} is driven by
both a synchronous and asynchronous type; last connect semantics will not

> determine the final type.

Reviewer 2: Everything before the `;` is restating enumeration item (2). Consider removing.

(synchronous reset) or `AsyncReset`{.firrtl} (asynchronous reset) during
compilation.

Synchronous resets used in registers will be mapped to a hardware description
language representation for synchronous resets.

The following example shows an uninferred reset that will get inferred to a
synchronous reset.

``` firrtl
input a : UInt<1>
wire reset : Reset
reset <= a
```

After reset inference, `reset` is inferred to the synchronous
`UInt<1>`{.firrtl} type:

``` firrtl
input a : UInt<1>
wire reset : UInt<1>
reset <= a
```

Asynchronous resets used in registers will be mapped to a hardware description
language representation for asynchronous resets.

The following example demonstrates usage of an asynchronous reset.

``` firrtl
input clock : Clock
input reset : AsyncReset
input x : UInt<8>
reg y : UInt<8>, clock with : (reset => (reset, UInt(123)))
; ...
```

Inference rules are as follows:

1. An uninferred reset driven by and/or driving only asynchronous resets will be
inferred as asynchronous reset.
1. An uninferred reset driven by and/or driving both asynchronous and synchronous
resets will cause an exception to be thrown. This occurs because reset inference

> happens before last-connect semantics are resolved.

Author: @youngar ty for the info on when reset inference happens; I've added it here!

Clock types are equivalent to clock types, and are not equivalent to any other
type.

An uninferred `Reset`{.firrtl} can be connected to either `UInt<1>`{.firrtl} or
`AsyncReset`{.firrtl}, but not both.

The `AsyncReset`{.firrtl} type is not equivalent to any other type.

> 

Author: @youngar current attempt at talking about reset equivalence 😬

```

Optionally, for the purposes of circuit initialization, a register can be

> declared with a reset signal (synchronous or asynchronous) and value. In the

Reviewer: ```suggestion
declared with a reset signal (uninferred, synchronous, or asynchronous) and value. In the
```

## Interpret as Clock

| Name    | Arguments | Parameters | Arg Types    | Result Type | Result Width |
|---------|-----------|------------|--------------|-------------|--------------|
| asClock | \(e\)     | ()         | (UInt)       | Clock       | n/a          |
|         |           |            | (SInt)       | Clock       | n/a          |
|         |           |            | (Fixed)      | Clock       | n/a          |
|         |           |            | (Clock)      | Clock       | n/a          |
|         |           |            | (AsyncReset) | Clock       | n/a          |

The result of the interpret as clock operation is the Clock typed signal
obtained from interpreting a single bit integer as a clock signal.

## Interpret as AsyncReset

| Name         | Arguments | Parameters | Arg Types    | Result Type | Result Width |
|--------------|-----------|------------|--------------|-------------|--------------|

> | asAsyncReset | \(e\)     | ()         | (AsyncReset) | AsyncReset  | n/a          |

Reviewer: ```suggestion
| asAsyncReset | \(e\)     | ()         | (AsyncReset) | AsyncReset  | n/a          |
| asAsyncReset |             |          | (Reset) | AsyncReset  | n/a          |
```

add Reset everywhere

(* Type definitions *)
width = "<" , int , ">" ;
binarypoint = "<<" , int , ">>" ;

> type_ground = "Clock" | "AsyncReset"

Reviewer: Where is Reset (uninferred) in this table?

Reviewer: Also we need to be careful if we just stick `Reset` in here, because this is used in extmodule.port.type ... is it legal to have a `Reset` here for extmodule? I'm not sure because Chisel doesn't even let you generate that.

> type_ground = "Clock" | "AsyncReset"

Reviewer: ```suggestion
type_ground = "Clock" | "AsyncReset" | "Reset"
```

Author: ```suggestion
type_ground = "Clock" | "AsyncReset"
type_abstract = "Reset"
...
// have a special extmodule port type
```

(* Type definitions *)
width = "<" , int , ">" ;
binarypoint = "<<" , int , ">>" ;

> type_ground = "Clock" | "AsyncReset"

Author: @seldridge I'm not sure how this is supposed to look 😬

           { port , newline } ,
           { statement , newline } ,
         dedent ;
port_extmodule = ( "input" | "output" ) , id , ":": , type_ground | type_aggregate , [ info ] ;
extmodule = "extmodule" , id , ":" , [ info ] , newline , indent ,

>               { port_extmodule , newline } ,

Author: @seldridge I think this is a bit messy, but you said you wanted to clean up the types later, so not sure if this flies. The notable thing that @mwachs5 and I discovered yesterday is that extmodules can't have uninferred `Reset`-type ports.

Reviewer: Ok, if we are not talking about this here we should mention it in the extmodule section I guess.

Overall I would desire that we could have inferred Resets (both inputs and outputs) on extmodule since it simplifies things  -- the other rules should kick in to interpret and enforce the inference rules. Such as if I have a Reset coming out of a black box and it drives an AsyncReset, the other rules already say we should know that extmodule's output is AsyncReset

Reviewer: ok great. Let's just put a bit in the `extmodule` section then about that?

Reviewer 2: So, technically that test is wrong. SFC allows uninferred resets on external module inputs, but not outputs. The test should be split into one test of input and one test of output.

I am actually not sure that this restriction has to be made. I.e., the reset inference algorithm should always do the right thing here.

Author: hmmm does MFC support resets on extmodule ports? I guess we can leave this stuff out if so.

Reviewer 2: For the following test:

Author: Cool, I'll just not say anything special about extmods then, since that makes the spec simpler, too!

           { port , newline } ,
           { statement , newline } ,
         dedent ;

> port_extmodule = ( "input" | "output" ) , id , ":": , type_ground | type_aggregate , [ info ] ;

Reviewer 2: This type of change blurs the line between syntax and semantics while also making the EBNF grammar more confusing. I.e., you likely do not have to make this differentiation because the EBNF grammar is only expressing legal syntax. If that fails type checking / type inference rules later, that's fine.

A good litmus test here is, "Should a FIRRTL _parser_ reject this code?" I would argue no.

I would not make the distinction between `port` and `port_extmodule`.

Author: kk, I will add `"Reset"` into `type_ground` then as well to simplify
